src/components/lib/cos_kernel_api.c

```diff
@@ -743,6 +827,50 @@ cos_page_bump_allocn(struct cos_compinfo *ci, size_t sz)
 	return (void *)__page_bump_alloc(ci, sz);
 }
 
+/* Allocates 4k pages backed by superpages or explicitly mapped superpages.
+ * Note that any range of explicitly mapped superpages will begin at the next 4MB
+ * interval, so the rest of the memory in the current 4MB space will be virtually
+ * inaccessible. Careful not to waste memory!
+ *   */
+void *
+cos_superpage_bump_allocn(struct cos_compinfo *ci, size_t sz, int superpage_aligned)
+{
+	int order;
+	vaddr_t heap_vaddr, heap_cursor, heap_limit;
+
+	order = (superpage_aligned) ? SUPER_PAGE_ORDER : PAGE_ORDER;
+	assert (sz % (1 << order) == 0);
+
+	/*
+	 * Allocate the virtual address range to map into.  This is
+	 * atomic, so we will get a contiguous range of sz.
+	 */
+	if (superpage_aligned) {
+		heap_vaddr = __superpage_bump_valloc(ci, sz);
+	} else {
+		heap_vaddr = __page_bump_valloc(ci, sz);
+	}
+	if (unlikely(!heap_vaddr)) return 0;
+	heap_limit = heap_vaddr + sz;
+	assert(heap_limit > heap_vaddr);
+
+	return (void *)__superpage_mem_bump_allocn(ci, sz, heap_vaddr, order);
+}
+
+void
+cos_retype_all_superpages(struct cos_compinfo *ci)
```

Comment:
I cannot imagine we want this function in this abstraction layer.

Further, relying on this function is preventing you from testing other more nuanced cases.  I'm not at all convinced that if this function is how you're testing the retype code that it actually works.